File: src/main/java/isop/api/ActionControllerExpression.java
package isop.api;

import isop.Build;
import isop.command_line.ControllerRecognizer;
import isop.command_line.parse.ArgumentParser;
import isop.command_line.parse.ParsedArguments;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.Map;
import java.util.function.Supplier;

/**
 * Created by mathieu.
 */
public class ActionControllerExpression {

    private final String controllerName;
    private final String actionName;
    private final Build build;

    public ActionControllerExpression(String controllerName, String actionName, Build build) {
        this.controllerName = controllerName;
        this.actionName = actionName;
        this.build = build;
    }
    public ParsedArguments parameters(Map<String, String> arg){
        ArgumentParser argumentParser = new ArgumentParser(build.getGlobalParameters(),
                build._allowInferParameter);
        ParsedArguments parsedArguments = argumentParser.parse(arg);
        ControllerRecognizer controllerRecognizer = build.getControllerRecognizers()
                .values().stream()
                .map(Supplier::get)
                .filter(this::isRecognized)
                .findFirst()
                .orElse(null);
        if (null != controllerRecognizer){
            return controllerRecognizer.parseArgumentsAndMerge(actionName, arg,
                    parsedArguments);
        }

        parsedArguments.assertFailOnUnMatched();
        return parsedArguments;
    }

    private boolean isRecognized(ControllerRecognizer recognizer) {
        return recognizer.recognize(new String[]{controllerName, actionName});
    }

    /// <summary>
    ///
    /// </summary>
    public String help(){
        throw new NotImplementedException();
    }
}

File: src/main/java/isop/api/ActionControllerExpression.kt
package isop.api

import isop.AppHostBuilder
import isop.abstractions.NotImplementedException
import isop.command_line.ControllerRecognizer
import isop.command_line.parse.ArgumentParser
import isop.command_line.parse.ParsedArguments

/**
 * Created by mathieu.
 */
class ActionControllerExpression(private val controllerName: String, private val actionName: String, private val build: AppHostBuilder) {
    fun parameters(arg: Map<String, String>): ParsedArguments {
        val argumentParser = ArgumentParser(build.globalParameters,
                build._allowInferParameter)
        val parsedArguments = argumentParser.parse(arg)
        val controllerRecognizer = build.controllerRecognizers
                .map { it.value() }
                .find { this.isRecognized(it) }
        if (null != controllerRecognizer) {
            return controllerRecognizer!!.parseArgumentsAndMerge(actionName, arg,
                    parsedArguments)
        }

        parsedArguments.assertFailOnUnMatched()
        return parsedArguments
    }

    private fun isRecognized(recognizer: ControllerRecognizer): Boolean {
        return recognizer.recognize(arrayOf(controllerName, actionName))
    }

    fun help(): String {
        throw NotImplementedException()
    }
}
